hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/IncrementalChunkBuffer.java

```diff
@@ -0,0 +1,274 @@
+        () -> this + ": Unexpected " + name + " at index " + i);
+  }
+
+  /** @return the i-th buffer if it exists; otherwise, return null. */
+  private ByteBuffer getAtIndex(int i) {
+    final ByteBuffer ith = i < buffers.size() ? buffers.get(i) : null;
+    if (ith != null) {
+      // assert limit/capacity
+      if (!isDuplicated) {
+        assertInt(getBufferCapacityAtIndex(i), ith.capacity(), "capacity", i);
+      } else {
+        final int limitIndex = limit / increment;
+        if (i < limitIndex) {
+          assertInt(increment, ith.capacity(), "capacity", i);
+        } else if (i == limitIndex) {
+          assertInt(getBufferCapacityAtIndex(i), ith.limit(), "capacity", i);
+        } else {
+          assertInt(0, ith.limit(), "capacity", i);
+        }
+      }
+    }
+    return ith;
+  }
+
+  /** @return the i-th buffer. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtIndex(int index) {
+    final int limitIndex = limit / increment;
+    // never allocate over limit
+    if (limit % increment == 0) {
+      Preconditions.checkArgument(index < limitIndex);
+    } else {
+      Preconditions.checkArgument(index <= limitIndex);
+    }
+
+    int size = buffers.size();
+    if (index < size) {
+      return getAtIndex(index);
+    }
+
+    // allocate upto the given index
+    ByteBuffer b = null;
+    for (; size <= index; size++) {
+      b = ByteBuffer.allocate(getBufferCapacityAtIndex(size));
+      buffers.add(b);
+    }
+    return b;
+  }
+
+  /** @return the buffer containing the position. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtPosition(int position) {
+    Preconditions.checkArgument(position < limit);
+    final int i = position / increment;
+    final ByteBuffer ith = getAndAllocateAtIndex(i);
+    assertInt(position%increment, ith.position(), "position", i);
+    return ith;
+  }
+
+  @Override
+  public int position() {
+    // The buffers list must be in the following orders:
+    // full buffers, buffer containing the position, empty buffers, null buffers
+    int i = 0;
+    // buffers must be full
+    for (; i < buffers.size(); i++) {
+      if (getAtIndex(i).position() != increment) {
+        break; // found the first non-full buffer
+      }
+    }
+    // buffer containing the position
+    final ByteBuffer ith = getAtIndex(i);
+    final int position = i * increment + Optional.ofNullable(ith)
+        .map(ByteBuffer::position).orElse(0);
+    // remaining buffers must be empty
+    assert assertRemainingList(ith, i);
+    return position;
+  }
+
+  private boolean assertRemainingList(ByteBuffer ith, int i) {
+    if (ith != null) {
+      // buffers must be empty
+      for (i++; i < buffers.size(); i++) {
+        ith = getAtIndex(i);
+        if (ith == null) {
+          break; // found the first non-null
+        }
+        assertInt(0, ith.position(), "position", i);
+      }
+    }
+    // buffers must be null
+    for (i++; i < buffers.size(); i++) {
```

Comment:
can you elaborate on the scenarios for line 157-160? my understanding is i will not be greater than buffers.size() in this for loop and getAtIndex(i) will not give us null in this case.

Reply:
You are right that, in the current implementation, we won't add null to the list so that getAtIndex(i) will never return null when I < buffers.size().  This is an assertion designed for a more general case earlier, where buffers(i) can be null (null is for unallocated buffers).

Let me replace it with asserting i == buffers.size().